keps/sig-storage/20190120-execution-hook-design.md

```diff
@@ -0,0 +1,546 @@
+        Spec ExecutionHookSpec
+
+        // Status defines the current state of a hook.
+        // +optional
+        Status ExecutionHookStatus
+}
+```
+
+Here is the definition of the ExecutionHookSpec:
+
+```
+// HookActionName is copied to ExecutionHookSpec by the controller such as
+// the Snapshot Controller.
+type ExecutionHookSpec struct {
+        // PodSelection defines how to select pods and containers to run
+	// the executionhook. This field is required.
+        PodSelection PodSelection
+
+        // Name of the HookAction. This is required.
+        HookActionName string
+}
+
+// PodSelection contains two fields, PodContainerNamesList and PodContainerSelector,
+// where one of them must be defined so that the hook controller knows where to
+// run the hook.
+type PodSelection struct {
+        // PodContainerNamesList lists the pods/containers on which the ExecutionHook
+        // should be executed. If not specified, the ExecutionHook controller will find
+        // all pods and containers based on PodContainerSelector.
+        // If both PodContainerNamesList and PodContainerSelector are not
+        // specified, the ExecutionHook cannot be executed and it will fail.
+        // +optional
+        PodContainerNamesList []PodContainerNames
+
+        // PodContainerSelector is for hook controller to find pods and containers
+        // based on the pod label selector and container names
+        // If PodContainerNamesList is specified, this field will not be used.
+        // +optional
+        PodContainerSelector *PodContainerSelector
+}
+
+type PodContainerNames struct {
+        PodName string
+        ContainerNames []string
+}
+
+type PodContainerSelector struct {
+	// PodSelector specifies a label query over a set of pods.
+        // +optional
+	PodSelector *metav1.LabelSelector
+
+        // ContainerSelector is a selector which must be true for the hook to run on the container
+        // Hook controller will find all containers in the selected pods based on ContainerSelector,
+        // if specified.
+        // The ContainerSeletor maps pod label to container name(s) so that the controller only select
+	// the containers that are listed in the map. Otherwise, all containers of the pods that match
+	// the pod label selector will be selected.
+        // +optional
+        ContainerSelector map[string]string
+ }
+```
+
+Here is the definition of the ExecutionHookStatus. This represents the current state of a hook for all selected containers in the selected pods.
+
+```
+// ExecutionHookStatus represents the current state of a hook
+type ExecutionHookStatus struct {
+        // This is a list of ContainerExecutionHookStatus, with each status representing information
+        // about how hook is executed in a container, including pod name, container name, ActionTimestamp,
+        // ActionSucceed, etc.
+        // +optional
+        ContainerExecutionHookStatuses []ContainerExecutionHookStatus
```

Comment:
Maybe structure this as `pods[] { name, containers[] { name, status { timestamp, success } } }` or `actions[] { podName, containerName, status { timestamp, success } }`

Reply:
In each ExecutionHook, we only define one action to be executed against one or many containers. So your second suggestion is basically the same as the one we use now, just the name is different?